File: Hard/Boolean Parenthesization/boolean-parenthesization.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution{
public:
const int mod=1003;
int solve(string &s,int i,int j,vector<vector<vector<int>>>&dp,int istrue)
{
    if(i>j)
    {
        return 0;
    }
    if(dp[i][j][istrue]!=-1)
    {
        return dp[i][j][istrue];
    }
    if(i==j)
    {
       if(istrue)
       {
           return s[i]=='T' ;
           
       }
       else
       {
           return s[i]=='F' ;
       }
    }
    int count=0;
    for(int idx=i+1;idx<j;idx+=2)
    {
        int lt=solve(s,i,idx-1,dp,1)%mod;
        int lf=solve(s,i,idx-1,dp,0)%mod;
        int rt=solve(s,idx+1,j,dp,1)%mod;
        int rf=solve(s,idx+1,j,dp,0)%mod;
        if(s[idx]=='|')
        {   
            if(istrue)
            {
            count=(count%mod+(lt*rf)%mod +(lf*rt)%mod+(lt*rt)%mod)%mod;
            
             }
            else
            {
            count=(count%mod+(lf*rf)%mod)%mod;
            }
        }
        if(s[idx]=='&')
        {
            if(istrue)
            {
              count=(count%mod+(lt*rt)%mod)%mod;  
            }
            else
            {
                count=(count%mod+(lf*rt)%mod+(lt*rf)%mod+(lf*rf)%mod)%mod;
            }
        }
        if(s[idx]=='^')
        {
            if(istrue)
            {
                count=(count%mod+(lf*rt)%mod+(rf*lt)%mod)%mod;
            }
            else
            {
                count=(count%mod+(lt*rt)%mod+(lf*rf)%mod)%mod;
            }
        }
        }
        return dp[i][j][istrue]=count%mod;
    }

    int countWays(int n, string s){
        // code here
        vector<vector<vector<int>>>dp(n+1,vector<vector<int>>(n+1,vector<int>(2,-1)));
        return solve(s,0,n-1,dp,1);
    }
};

//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        string s;
        cin>>s;
        
        Solution ob;
        cout<<ob.countWays(n, s)<<"\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Boolean Parenthesization/boolean-parenthesization.py
#User function Template for python3
class Solution:
    def evaluate(self, b1, b2, op):
        if op == '&':
            return (b1 & b2) == 1
        if op == '|':
            return (b1 | b2) == 1
        return (b1 ^ b2) == 1

    def countWays(self, s):
        n = len(s)
        dp = [[[0, 0] for _ in range(n)] for _ in range(n)]

        for i in range(0, n, 2):
            dp[i][i][1] = 1 if s[i] == 'T' else 0
            dp[i][i][0] = 1 if s[i] == 'F' else 0

        for length in range(2, n, 2):
            for i in range(0, n - length, 2):
                j = i + length
                dp[i][j][0] = dp[i][j][1] = 0

                for k in range(i + 1, j, 2):
                    op = s[k]
                    leftTrue = dp[i][k - 1][1]
                    leftFalse = dp[i][k - 1][0]
                    rightTrue = dp[k + 1][j][1]
                    rightFalse = dp[k + 1][j][0]

                    if self.evaluate(1, 1, op):
                        dp[i][j][1] += leftTrue * rightTrue
                    if self.evaluate(1, 0, op):
                        dp[i][j][1] += leftTrue * rightFalse
                    if self.evaluate(0, 1, op):
                        dp[i][j][1] += leftFalse * rightTrue
                    if self.evaluate(0, 0, op):
                        dp[i][j][1] += leftFalse * rightFalse

                    if not self.evaluate(1, 1, op):
                        dp[i][j][0] += leftTrue * rightTrue
                    if not self.evaluate(1, 0, op):
                        dp[i][j][0] += leftTrue * rightFalse
                    if not self.evaluate(0, 1, op):
                        dp[i][j][0] += leftFalse * rightTrue
                    if not self.evaluate(0, 0, op):
                        dp[i][j][0] += leftFalse * rightFalse

        return dp[0][n - 1][1]  # code here


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        s = input().strip()
        print(Solution().countWays(s))
        print("~")

# } Driver Code Ends